### ISSUE TITLE
Explicit selfdestruct

### ISSUE THREAD

**Issue by pepyakin** [paritytech/substrate#3254]
_Originally posted by @pepyakin in https://github.com/paritytech/substrate/issues/2705#issuecomment-499888672_

After #3255, we could implement another deferred action - inevitabile deliberate self destruction. It would not have a problem that the current "drain-to-death" mechanism has, i.e. it won't be possible to send funds to a contract to prevent its self destruction.

**Comment by DemiMarie-parity** [comment]
I recommend against this, because of the multi-sig wallet bug.

**Comment by pepyakin** [comment]
I believe this particular bug doesn't mean that we should reject the idea of selfdestruct. Especially considering that the problem is not caused by selfdestruct exclusively but rather because of combination of factors like using a (selfdestructable) contract as a dynamic library.

I think selfdestruct is useful on its own since there are cases when you really done with a contract and there is nothing better to do that just remove it. Think of an agreement-like contract, when the contract performed the conditional logic it was created for.

srml-contracts, unlike Ethereum, is not a homogenious single piece but is a moudle and a part of SRML and it has to integrate into it. One conern that it has to respect is the way that the outer system handles accounting. In SRML if an account's balance falls below certain threshold the account is removed. Because a contract is also an account it is removed when it ends up with its balance below such treshold. We kind of already have a way to selfdestruct, dubbed, "drain-to-death".

> This is not to mention state rent which purpose is to kill accounts which can't pay for their storage.

However, "drain-to-death" has an issue - after the contract sent all of its funds it is possible that another contract sends funds to that contract until the end of transaction effectively blocking the self destruction. Adding an explicit request for termination can fix that.

Btw, srml-contracts separates code from a contract paving the way for the first-class dynamic libraries (This is not to say that it is completely eliminates the risk that somebody decide to delegate some critical logic to a selfdestructable contract, but rather that this feature would make that less probable)

**Comment by andresilva** [comment]
What happens if we selfdestruct a contract with funds? Would this be an effective way for burning funds? Otherwise if we don't want to allow that, we can make the selfdestruct only work when balance is below existential. This should still prevent the current problem of sending funds to contract to defer destruction.

**Comment by jimpo** [comment]
This wouldn't be a problem if contract destruction is not deferred. I propose the following:

- If a contract is called and returns from a call with a balance below the existential deposit, the contract is immediately removed and its balance is set to 0. This is visible to any contracts calling it higher in the execution stack.
  - There is an exception if the contract with insufficient balance is being recursively called from itself. If a contract returns with balance below existential deposit and is also being executed higher up in the execution stack, then the nested call traps immediately. The caller sees the trap and may be able to recover.
- If a contract is instantiated and returns from the constructor with a balance below the existential deposit, then this traps immediately. The caller sees the trap and may be able to recover.
- If a contract self-destructs, then the caller sends it a balance exceeding the existential deposit, then this creates a new account at that address with no contract code.
- If a contract self-destructs, then a caller may still deploy the same contract with the same code hash and input data such that the same address is assigned. This replaces the old contract with the newly instantiated contract. (Exactly what would happen if this occurred in a subsequent extrinsic).

The benefits of this approach is that it ensures a contract cannot be destroyed with a remaining balance and it seems more like expected behavior to me that a contract removal is applied immediately.

**Comment by jimpo** [comment]
Closed by #3417.
